lib/runtime-core/src/memory/mod.rs

```diff
@@ -289,27 +282,56 @@ impl Clone for UnsharedMemory {
 }
 
 pub struct SharedMemory {
-    #[allow(dead_code)]
-    desc: MemoryDescriptor,
+    internal: Arc<SharedMemoryInternal>,
```

Comment:
Do we need an internal structure? can we move all the fields from `SharedMemoryInternal ` into the `SharedMemory` itself?

Reply:
The issue is the implementation of Clone for SharedMemory. Mind you, I'm suspicious of the fact that we clone memories to begin with, but I took a look and decided that was a problem for another day. The important part is that the underlying size/grow methods share one lock, which is held by `SharedMemoryInternal` while the `SharedMemory` can be cloned to your heart's content.